File: mcp/src/client/network.ts
// ─────────────────────────────────────────────────────────────────────────────
// network.ts — the two-chain routing table. THE single source of truth for
// which chain each transaction-builder endpoint targets. A V2 tx is built
// against one chain's blockhash and MUST be submitted to that chain; mixing
// them fails to land. Every tx tool reads this map; the routing unit tests
// assert it. If a new builder endpoint is added, add it here or the build breaks.
// ─────────────────────────────────────────────────────────────────────────────

/** Which chain a built transaction must be submitted to. */
export type Chain = 'er' | 'base'

/**
 * Endpoint class → chain. Keys are the transaction-builder path segments.
 * - `er`   → Ephemeral Rollup (trading): open/close/collateral/triggers/limits
 * - `base` → base-chain Solana (setup + withdrawal)
 * Source: examples-v2 lifecycle.ts chain map + client.ts per-method TSDoc.
 */
export const TX_CHAIN = {
  // Trading — Ephemeral Rollup
  'open-position': 'er',
  'close-position': 'er',
  'reverse-position': 'er',
  'add-collateral': 'er',
  'remove-collateral': 'er',
  'place-tp-sl': 'er',
  'place-trigger-order': 'er',
  'edit-trigger-order': 'er',
  'cancel-trigger-order': 'er',
  'cancel-all-trigger-orders': 'er',
  'edit-limit-order': 'er',
  'cancel-limit-order': 'er',
  // Account setup — base chain
  'init-basket': 'base',
  'init-deposit-ledger': 'base',
  'delegate-basket': 'base',
  'deposit-direct': 'base',
  // Withdrawal — base chain
  'request-withdrawal': 'base',
  'execute-withdrawal': 'base',
} as const satisfies Record<string, Chain>

export type TxKind = keyof typeof TX_CHAIN

/** The chain a given builder endpoint targets. */
export function chainForTx(kind: TxKind): Chain {
  return TX_CHAIN[kind]
}

/** Human note appended to tool output so the caller passes the right value to sign_and_send. */
export function chainHint(kind: TxKind): string {
  const chain = TX_CHAIN[kind]
  return chain === 'er'
    ? 'Submit on the Ephemeral Rollup: call sign_and_send with network="er".'
    : 'Submit on the base chain: call sign_and_send with network="base".'
}

File: cli/src/core/network.rs
// ─────────────────────────────────────────────────────────────────────────────
// network.rs — the V2 two-chain model. Trading txs submit to the Ephemeral
// Rollup RPC; setup + withdrawal txs submit to the base-chain Solana RPC.
// The API base serves the V2 surface at ROOT (the /v2 edge prefix is not
// deployed). Env overrides: FLASH_API_URL, ER_RPC_URL, SOLANA_RPC_URL.
// ─────────────────────────────────────────────────────────────────────────────

/// Which chain a built transaction must be submitted to.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Chain {
    /// Ephemeral Rollup — trading transactions.
    Er,
    /// Base-chain Solana — setup + withdrawal transactions.
    Base,
}

impl Chain {
    pub fn label(self) -> &'static str {
        match self {
            Chain::Er => "Ephemeral Rollup",
            Chain::Base => "base chain",
        }
    }
}

#[derive(Debug, Clone)]
pub struct Network {
    /// Hosted V2 REST base (root, no trailing slash).
    pub api_base: String,
    /// Ephemeral Rollup RPC — trading tx submission.
    pub er_rpc: String,
    /// Base-chain Solana RPC — setup + withdrawal tx submission.
    pub base_rpc: String,
}

impl Default for Network {
    fn default() -> Self {
        Network {
            api_base: "https://flashapi.trade".to_string(),
            er_rpc: "https://flash.magicblock.xyz".to_string(),
            base_rpc: "https://api.mainnet-beta.solana.com".to_string(),
        }
    }
}

impl Network {
    /// Resolve from the environment, falling back to verified defaults.
    pub fn resolve() -> Self {
        let d = Network::default();
        Network {
            api_base: std::env::var("FLASH_API_URL")
                .map(|s| s.trim_end_matches('/').to_string())
                .unwrap_or(d.api_base),
            er_rpc: std::env::var("ER_RPC_URL").unwrap_or(d.er_rpc),
            base_rpc: std::env::var("SOLANA_RPC_URL").unwrap_or(d.base_rpc),
        }
    }

    /// The RPC endpoint for a given chain.
    pub fn rpc_for(&self, chain: Chain) -> &str {
        match chain {
            Chain::Er => &self.er_rpc,
            Chain::Base => &self.base_rpc,
        }
    }
}

/// The routing table: which chain each transaction-builder endpoint targets.
/// Single source of truth; the routing test asserts it.
pub fn chain_for(builder_path: &str) -> Chain {
    match builder_path {
        // Trading — Ephemeral Rollup
        "open-position" | "close-position" | "reverse-position" | "add-collateral"
        | "remove-collateral" | "place-tp-sl" | "place-trigger-order" | "edit-trigger-order"
        | "cancel-trigger-order" | "cancel-all-trigger-orders" | "edit-limit-order"
        | "cancel-limit-order" => Chain::Er,
        // Account setup + withdrawal — base chain
        "init-basket" | "init-deposit-ledger" | "delegate-basket" | "deposit-direct"
        | "request-withdrawal" | "execute-withdrawal" => Chain::Base,
        // Default to base for anything unrecognized (safest for setup-like ops).
        _ => Chain::Base,
    }
}
